vault/identity_store_oidc.go

```diff
@@ -0,0 +1,743 @@
+			}
+			overrides["verificationttlDuration"] = verificationttlDuration
+		}
+
+		// load the named key and perform a rotation
+		entry, err := req.Storage.Get(ctx, "oidc-config/named-key/"+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry == nil {
+			return logical.ErrorResponse(fmt.Sprintf("no named key found at %q", name)), logical.ErrInvalidRequest
+		}
+
+		var storedNamedKey NamedKey
+		if err := entry.DecodeJSON(&storedNamedKey); err != nil {
+			return nil, err
+		}
+		err = storedNamedKey.rotate(overrides)
+		if err != nil {
+			return nil, err
+		}
+
+		// prepare response
+		if verificationttlInput != "" {
+			responseVerificationttl = verificationttlInput
+		} else {
+			responseVerificationttl = storedNamedKey.Verificationttl
+		}
+
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"rotation_period":  storedNamedKey.RotationPeriod,
+				"verification_ttl": responseVerificationttl,
+				"algorithm":        storedNamedKey.SigningAlgorithm.String(),
+			},
+		}, nil
+	}
+}
+
+// handleOIDCReadCerts is used to retrieve all certs from all keys so that clients can
+// verify the validity of a signed OIDC token
+func (i *IdentityStore) handleOIDCReadCerts() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+
+		jwks := jose.JSONWebKeySet{
+			Keys: make([]jose.JSONWebKey, len(publicKeys)),
+		}
+
+		for i, expirableKey := range publicKeys {
+			jwks.Keys[i] = expirableKey.Key
+		}
+
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"keys": jwks,
+			},
+		}, nil
+	}
+}
+
+// --- some helper methods ---
+
+// NamedKey.rotate(overrides) performs a key rotation on a NamedKey
+// ASSUMPTION: overrides make sense and won't cause type conversion errors...
+func (namedKey NamedKey) rotate(overrides map[string]interface{}) error {
+	// fail early if the verification ttl of the named key is not parse-able
+	var verificationttlDuration time.Duration
+	verificationttlDurationInterface, ok := overrides["verificationttlDuration"]
+	if ok {
+		verificationttlDuration = verificationttlDurationInterface.(time.Duration)
+	} else {
+		var err error
+		verificationttlDuration, err = parseutil.ParseDurationSecond(namedKey.Verificationttl)
+		if err != nil {
+			return err
+		}
+	}
+
+	// generate new key
+	signingKey, publicKey, err := generateKeys(namedKey.SigningAlgorithm)
+	if err != nil {
+		return err
+	}
+
+	// rotation involves overwritting current signing key, updating key ring, and updating global
+	// public keys to expire the signing key that was just rotated
+	rotateID := namedKey.SigningKey.KeyID
+	namedKey.SigningKey = signingKey
+	publicKeys = append(publicKeys, publicKey)
+	namedKey.KeyRing = append(namedKey.KeyRing, publicKey.Key.KeyID)
+
+	// give current signing key's public portion an expiry time
+	for _, publicKey := range publicKeys {
```

Comment:
Maybe choose another name instead of `publicKey`. I got a little confused skimming this since that's use earlier for the new key.